File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy = prices[0]
        maxProfit = 0
        
        for right in range(1,len(prices)):
            if prices[right] < buy:
                buy = prices[right]
            else:
                profit = prices[right] - buy
                maxProfit = max(maxProfit,profit)
        
        return maxProfit
        

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int buy_price = prices[0];
        int profit = 0;//minimum profit is zero
        
        for(int i = 1; i < prices.length; i++) {
            if(prices[i] < buy_price) { // take the smallest buy price
                buy_price = prices[i];
            } else {
                int current_profit = prices[i] - buy_price;//sell - buy
                profit = Math.max(profit,current_profit);
            }
        }
        
        return profit;
    }
}